Refactor metrics and  make the cacheManager more generic

zyfjeff opened the pull request:
1. enable self-contained metrics tracking for cacheManager, without relying on cachedStore for tracking
2. separating metrics from the cacheManager allows other object to use them.
3. remove the metrics "stageBlockDelay" from the cacheManager and place it in the cachedStore.

zyfjeff commented:
The stageBlockDelay metric is always used in cachedStore, so I don’t think it is a metric that belongs to cacheManager. This causes cachedStore to be unable to use CacheManager through interface, and always needs to convert it to cacheManager, and then get metrics

zyfjeff commented:
memstore and diskstore has some common metrics, so I separate metrics from the cacheManager and make the memstore can use it

zyfjeff commented:
ping @zhijian-pro

zyfjeff commented:
@SandyXSD Are there any comments?

SandyXSD commented on pkg/chunk/disk_cache.go:
```diff
@@ -156,6 +156,18 @@ func (cache *cacheStore) refreshCacheKeys() {
+			cache.m.stageBlockBytes.Sub(float64(size))
+		}
+	}
+	return err
```
ignore the ENOENT error

zyfjeff replied:
fixed

SandyXSD commented on pkg/chunk/utils_linux.go:
```diff
@@ -28,3 +32,43 @@ func getAtime(fi os.FileInfo) time.Time {
 }
+
+// Copy from https://github.com/prometheus/client_golang/blob/v1.14.0/prometheus/testutil/testutil.go
+func ToFloat64(c prometheus.Collector) float64 {
```
move this func to the test file.

zyfjeff replied:
fixed

SandyXSD commented on pkg/chunk/metrics.go:
```diff
@@ -0,0 +1,81 @@
+import "github.com/prometheus/client_golang/prometheus"
+
+// CacheManager Metrics
+type CacheManagerMetrics struct {
```
unexport it?

zyfjeff replied:
fixed

zyfjeff commented:
```
2023-04-10T06:37:57.3152971Z 2023/04/10 06:33:46.372653 juicefs[25700] <INFO>: Create session 1 OK with version: 1.1.0-dev+unknown [base.go:459]
2023-04-10T06:37:57.3153608Z 2023/04/10 06:33:46.555170 juicefs[25700] <INFO>: close session 1: %!s(<nil>) [base.go:572]
2023-04-10T06:37:57.3158325Z 2023/04/10 06:33:46.690422 juicefs[25700] <WARNING>: File name is too long as a trash entry, truncating it: fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff -> 1-21-ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff [base.go:1972]
2023-04-10T06:37:57.3162458Z 2023/04/10 06:33:46.701344 juicefs[25700] <INFO>: cleanup trash: deleted 6 files in 9.196286ms [base.go:2020]
2023-04-10T06:37:57.3163075Z 2023/04/10 06:33:49.063360 juicefs[25700] <WARNING>: checking counter lastCleanupFiles: read-only file system [base.go:809]
2023-04-10T06:37:57.3163634Z 2023/04/10 06:34:00.702454 juicefs[25700] <INFO>: chunk 4137:1 was changed 72 -> 24 [sql.go:2647]
2023-04-10T06:37:57.3165143Z 2023/04/10 06:34:00.702566 juicefs[25700] <INFO>: compacted chunk 1325 was not used [sql.go:2690]
2023-04-10T06:37:57.3165580Z 2023/04/10 06:34:00.702592 juicefs[25700] <INFO>: compaction for 4137:1 is wasted, delete slice 1325 (67108864 bytes) [sql.go:2697]
2023-04-10T06:37:57.3166205Z 2023/04/10 06:34:04.638843 juicefs[25700] <INFO>: Create session 2 OK with version: 1.1.0-dev+unknown [base.go:459]
2023-04-10T06:37:57.3166622Z 2023/04/10 06:34:04.649043 juicefs[25700] <INFO>: close session 2: %!s(<nil>) [base.go:572]
2023-04-10T06:37:57.3167170Z 2023/04/10 06:34:05.232718 juicefs[25700] <INFO>: Create session 3 OK with version: 1.1.0-dev+unknown [base.go:459]
2023-04-10T06:37:57.3167585Z 2023/04/10 06:34:10.405013 juicefs[25700] <INFO>: Quota for inode 4455 is deleted [base.go:597]
2023-04-10T06:37:57.3167968Z 2023/04/10 06:34:10.406931 juicefs[25700] <INFO>: close session 3: %!s(<nil>) [base.go:572]
2023-04-10T06:37:57.3168372Z 2023/04/10 06:34:10.885635 juicefs[25700] <WARNING>: nlink of /check should be 3, but got 0 [base.go:1746]
2023-04-10T06:37:57.3169052Z 2023/04/10 06:34:10.885668 juicefs[25700] <WARNING>: Path /check (inode 4467) can be repaired, please re-run with '--path /check --repair' to fix it [base.go:1773]
2023-04-10T06:37:57.3169523Z 2023/04/10 06:34:10.886145 juicefs[25700] <WARNING>: nlink of /check should be 3, but got 0 [base.go:1746]
2023-04-10T06:37:57.3169955Z 2023/04/10 06:34:10.889273 juicefs[25700] <WARNING>: nlink of /check/d1/d2 should be 3, but got 0 [base.go:1746]
2023-04-10T06:37:57.3170361Z 2023/04/10 06:34:10.891892 juicefs[25700] <WARNING>: attribute of / is missing [base.go:1750]
2023-04-10T06:37:57.3170772Z 2023/04/10 06:34:11.007206 juicefs[25700] <WARNING>: nlink of /check/d1 should be 3, but got 0 [base.go:1746]
2023-04-10T06:37:57.3171202Z 2023/04/10 06:34:11.012559 juicefs[25700] <WARNING>: nlink of /check/d1/d2/d3 should be 3, but got 0 [base.go:1746]
2023-04-10T06:37:57.3171638Z 2023/04/10 06:34:11.014496 juicefs[25700] <WARNING>: nlink of /check/d1/d2/d3/d4 should be 2, but got 0 [base.go:1746]
2023-04-10T06:37:57.3172212Z 2023/04/10 06:34:21.409117 juicefs[25700] <INFO>: Create session 1 OK with version: 1.1.0-dev+unknown [base.go:459]
2023-04-10T06:37:57.3172612Z 2023/04/10 06:34:21.409236 juicefs[25700] <INFO>: Quota for inode 4454 is deleted [base.go:597]
2023-04-10T06:37:57.3173004Z 2023/04/10 06:34:21.409255 juicefs[25700] <INFO>: Quota for inode 4458 is deleted [base.go:597]
2023-04-10T06:37:57.3173388Z     base_test.go:2211: added space: 65536
2023-04-10T06:37:57.3173729Z 2023/04/10 06:34:21.454137 juicefs[25700] <INFO>: close session 1: %!s(<nil>) [base.go:572]
2023-04-10T06:37:57.3174113Z --- FAIL: TestSQLiteClient (35.11s)
2023-04-10T06:37:57.3174393Z === RUN   TestMySQLClient
2023-04-10T06:37:57.3175023Z 2023/04/10 06:34:21.535337 juicefs[25700] <ERROR>: error: Error 1146: Table 'dev.jfs_node' doesn't exist
```
Flakey Test

zyfjeff commented:
/retest

zyfjeff commented:
@SandyXSD unittests passed
